### ISSUE TITLE
About decoding generated words.

### ISSUE THREAD

**Issue by q-rong** [hobincar/RecNet#22]
Hi,

I noticed some of the papers mentioned about generating words, and they were softmax on the `hidden` . However, your code is softmax for the `output` generated by LSTM.

I am very confused about this. I don't quite understand why you didn't use `hidden`?

```
decoder.py :

 def forward(self, input, hidden, feats):
        embedded = self.embedding(input)
        last_hidden = self.get_last_hidden(hidden)
        feats, attn_weights = self.attention(last_hidden, feats)
        input_combined = torch.cat((
            embedded,
            feats.unsqueeze(0)), dim=2)
        output, hidden = self.rnn(input_combined, hidden)
        output = output.squeeze(0)
        output = self.out(output)
        output = F.log_softmax(output, dim=1)
        return output, hidden, attn_weights
```

**Comment by hobincar** [comment]
The return value of `torch.nn.LSTM` is `output` and `hidden`. `output` contains the final hidden state of LSTM and `hidden` container all intermediate hidden states and context states. We can get the final hidden state from the `hidden`, but just using `output` is an easier way to do the same thing.

**Comment by q-rong** [comment]
Hi,

`caption_generator.py >> def beam_search( ) :`

In the `beam_search` function, I notice that `outputs` is the most desired. So, do the parameters associated with `hidden` still matter?

As I understand your code, you first get `next_hidden` through `self.decoder_2`, then `beam_hidden_list`, then `topk_hidden_list`, and finally `hidden_list`.
You use the resulting `hidden_list` for `assert len(input_list) == len(hidden_list) == len(cum_prob_list)`.

 Just like `beam_hidden_list`, `topk_hidden_list`, `hidden_list`, can we delete the content associated with `hidden`?
